File: solutions/Leetcode_105/leetcode_105.cc
# include <unordered_map>
# include <vector>

using namespace std;

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    {
        unordered_map<int, int> indices;
        for (int i = 0; i != inorder.size(); i++)
            indices[inorder[i]] = i;
        return buildSubTree(preorder, inorder, indices, 0, 0, inorder.size()-1);
    }

private:
    TreeNode *buildSubTree(vector<int>& preorder, vector<int> &inorder, unordered_map<int, int> &indices, int idx, int low, int high)
    {
        if (low > high)
            return nullptr;
        TreeNode *root = new TreeNode(preorder[idx]);
        int mid = indices[root->val];
        root->left = buildSubTree(preorder, inorder, indices, idx+1, low, mid-1);
        root->right = buildSubTree(preorder, inorder, indices, idx+mid-low+1, mid+1, high);
        return root;
    }
};

File: solutions/Leetcode_105/leetcode_105.py
from typing import List

class Solution1:
    def build_tree(self, preorder, inorder):
        indices = {x: i for i, x in enumerate(inorder)}

        def build_tree(idx, low, high):
            if low > high:
                return None
            root = TreeNode(preorder[idx])
            mid = indices[root.val]
            root.left = build_tree(idx+1, low, mid-1)
            root.right = build_tree(idx+mid-low+1, mid+1, high)
            return root
        
        return build_tree(0, 0, len(inorder)-1)

class SolutionMINE:
    def build_tree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder:
            return None
        value = preorder[0]
        root = TreeNode(value)
        idx = inorder.index(value)
        root.left = self.build_tree(preorder[:idx], inorder[:idx])
        root.right = self.build_tree(preorder[idx+1:], inorder[idx+1:])
        return root